[jit] make FunctionType a NamedType

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23799 [wip] serializing function calls
* #23846 [jit] clean up import_source
* #23845 [jit] class_table_ to deps_table_
* **#23697 [jit] make FunctionType a NamedType**
* #23696 [jit] make NamedType an interface
* #23691 [jit] simplify NamedType interface
* #23689 [jit] search class type for methods

This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
[jit] make FunctionType a NamedType

This simplifies the groundwork for serializing functions.

ZolotukhinM commented on torch/csrc/jit/script/class_type.cpp:
```diff
@@ -6,6 +6,10 @@ namespace c10 {
+FunctionType::FunctionType(Function* function)
+    : NamedType(TypeKind::FunctionType),
+      function_(function),
+      name_(function_->qualname()) {}
```
Note that while this code is correct, it is susceptible to the order of fields in `FunctionType`. If for some reason the code will get shuffled in future, we might call `function_->qualname()` before `function_` is initialized. I think it would be safer if we use `name_(function->qualname())` instead of `name_(function_->qualname())`.

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)

suo committed:
Update on "[jit] make FunctionType a NamedType"


This simplifies the groundwork for serializing functions.

Differential Revision: [D16611884](https://our.internmc.facebook.com/intern/diff/D16611884)
